text/0195-kubernetes-resource-detection.md

```diff
@@ -0,0 +1,113 @@
+# Environment Variable Resource Detection in Kubernetes
+
+This OTEP proposes to standardize environment variables used to detect resources exposed through the [Kubernetes downward API](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/) to provide a consistent experience on Kubernetes across languages and the collector.
+
+There is a mismatch between the format otel expects resources to be specified in for environment variables (a single, comma-deliniated list of key-value pairs in `OTEL_RESOURCE_ATTRIBUTES`), and the format Kubernetes exposes resource information in (one environment-variable for each attribute).  This makes correct resource detection in Kubernetes difficult, and the lack of specification has led to differing implementations across language SDKs and the collector.
+
+## Background
+
+Adding Kubernetes resources attributes to telemetry can be done one of three ways, depending on your telemetry collection pipeline:
+
+1. The application detects its own resource attributes from environment variables
+2. A "sidecar" collector container detects its own resource attributes from environment variables
+3. A centralized collector detects resource attributes for many Kubernetes workloads by querying the kube-apiserver.
+
+Method 3 is already implemented today as the `k8sattributesprocessor`, but requires running a collector, and adds additional load on the kube-apiserver. Methods 1 and 2 require substantial configuration today, which differs based on the language used, or if using a sidecar collector.
+
+Not all attributes in the [k8s semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/semantic_conventions/resource/k8s.yaml) are supported by the Kubernetes downward API.  The OpenTelemetry attributes that could be supported ([api reference](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#envvarsource-v1-core)) are:
```

Comment:
Although the container name isn't accessible from the downward API, it's still possible to

* standardize on an environment variable approach for injecting it
* inject this as an environment variable in a uniform way (e.g. mutating admission webhook)

Since this OTEP is scoped to "Env variables on Kubernetes" not just "Env variables via downward API", I think it's reasonable to cover it here.

Reply:
There are a bunch of fields I don't cover here, since I'm trying to stick to fields available through the downward API. 
I'm going to talk to some k8s folks to see what it would take to have some of those missing conventions added to the downward API's list of allowed fieldPaths.  Today, you can just set OTEL_RESOURCE_ATTRIBUTES to: `k8s.container.name=my-container-name`, but that obviously isn't a great solution.

Reply:
Note that it might be tricky to detect the container name in a sidecar: either it will inject the sidecar's name and not the workload's, or it will have to try to determine which container contains the actual workload. While this is possible in most cases, there are cases where pods have multiple sidecars (3+ containers in the pod), so that one sidecar can't reliably tell which of the remaining two is the actual workload.